osu.Game/Screens/OnlinePlay/OnlinePlaySongSelect.cs

```diff
@@ -155,6 +164,10 @@ protected override IEnumerable<(FooterButton, OverlayContainer)> CreateFooterBut
         /// </summary>
         /// <param name="mod">The <see cref="Mod"/> to check.</param>
         /// <returns>Whether <paramref name="mod"/> is a selectable free-mod.</returns>
-        protected virtual bool IsValidFreeMod(Mod mod) => IsValidMod(mod);
+        protected virtual bool IsValidFreeMod(Mod mod) => IsValidMod(mod) && checkCompatibleFreeMod(mod);
+
+        private bool checkCompatibleFreeMod(Mod mod)
+            => Mods.Value.All(m => m.Acronym != mod.Acronym) // Mod must not be contained in the required mods.
+               && ModUtils.CheckCompatibleSet(Mods.Value.Append(mod).ToArray()); // Mod must be compatible with all the required mods.
```

Comment:
This doesn't catch all cases. Multi mods are one that I have on my mind in particular; for example, you can select Perfect as a required mod and add Sudden Death as a free mod, which doesn't make much sense. Worse yet, there are crashing cases - e.g. in mania, selecting Hidden as required and Fade In as free results in a crash with the following stack trace:

```
[runtime] 2021-02-22 21:18:20 [error]: System.InvalidCastException: Unable to cast object of type 'osu.Framework.Graphics.Containers.BufferedContainer' to type 'osu.Framework.Graphics.Containers.Container'.
[runtime] 2021-02-22 21:18:20 [error]: at osu.Game.Rulesets.Mania.Mods.ManiaModHidden.ApplyToDrawableRuleset(DrawableRuleset`1 drawableRuleset) in /home/dachb/Dokumenty/opensource/osu/osu.Game.Rulesets.Mania/Mods/ManiaModHidden.cs:line 33
[runtime] 2021-02-22 21:18:20 [error]: at osu.Game.Rulesets.UI.DrawableRuleset`1.applyRulesetMods(IReadOnlyList`1 mods, OsuConfigManager config) in /home/dachb/Dokumenty/opensource/osu/osu.Game/Rulesets/UI/DrawableRuleset.cs:line 349
```

Not sure whether this should be handled in this check, or at the level of incompatibility rules, so this might not be in scope of this pull, but I figured I'd better point this out.

Reply:
Addressed in https://github.com/ppy/osu/pull/11891